Projects/Shapes/Shapes/Shapes/DynamicArray.h

```diff
@@ -0,0 +1,341 @@
+	const T& Front() const
+	{
+		return Array[0];
+	}
+
+	//-------------------------------------------------------------------------------------------
+	T& Back()
+	{
+		return Array[Size - 1];
+	}
+
+	//-------------------------------------------------------------------------------------------
+	const T& Back() const
+	{
+		return Array[Size - 1];
+	}
+
+	//-------------------------------------------------------------------------------------------
+	T* GetData()
+	{
+		return Array;
+	};
+
+	//-------------------------------------------------------------------------------------------
+	const T* GetData() const
+	{
+		return Array;
+	};
+
+	//-------------------------------------------------------------------------------------------
+	//											CAPACITY
+	//-------------------------------------------------------------------------------------------
+
+	bool IsEmpty() const
+	{
+		return Size == 0;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	int GetSize() const
+	{
+		return Size;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	int GetCapacity() const
+	{
+		return Capacity;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	bool Reserve(const int NewCapacity)
+	{
+		if (NewCapacity <= Capacity)
+		{
+			return false;
+		}
+
+		T* NewArray = new T[NewCapacity];
+		for (int i = 0; i < Size; i++)
+		{
+			NewArray[i] = Array[i];
+		}
+
+		delete[] Array;
+		Array = NewArray;
+
+		Capacity = NewCapacity;
+		return true;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	bool ShrinkToFit()
+	{
+		if (Capacity == Size)
+		{
+			return false;
+		}
+
+		T* NewArray = new T[Size];
+		for (int i = 0; i < Size; i++)
+		{
+			NewArray[i] = Array[i];
+		}
+
+		delete[] Array;
+		Array = NewArray;
+
+		Capacity = Size;
+		return true;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	//											MODIFIERS
+	//-------------------------------------------------------------------------------------------
+
+	void Clear()
+	{
+		Size = 0;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void PushBack(const T& Item)
+	{
+		if (Size == Capacity)
+		{
+			if (Capacity == 0)
+			{
+				Reserve(1);
+			}
+			Reserve(Capacity * 2);
+		}
+		Array[Size] = Item;
+		Size++;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void PopBack()
+	{
+		if (Size > 0)
+		{
+			Size--;
+		}
+	}
+
+	//-------------------------------------------------------------------------------------------
+	bool Insert(const T& Item, const int Index)
+	{
+		if (Index > Size)
+		{
+			return false;
+		}
+
+		if (Index == Size)
+		{
+			PushBack(Item);
+			Size++;
+			return true;
+		}
+
+		if (Size == Capacity)
```

Comment:
This one is dangerous considering that your initial capacity and initial size are both 0 if not specified otherwise.